In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

## Notebook Information and Motivation

### Table of Contents
1) Introduction to Exploration
2) Objectives & Questions
3) Data Loading
4) Dataset Information
5) Preliminary Explorations with Pandas

### Objectives & Questions

### Data Loading

In [3]:
df_gas = pd.read_excel("../data/raw/Gas_Prices.xls", sheet_name="Data 1")
df_EVs = pd.read_csv("../data/raw/Electric_Vehicle_Population_Data.csv")

For df_gas, I needed to use the second sheet, which is why I specified the sheet name as "Data 1". This is because the first sheet had contact information of the organization who made the dataset. 

### Dataset Information


For this analysis, I used two datasets:
https://catalog.data.gov/dataset/electric-vehicle-population-data
https://www.eia.gov/dnav/pet/hist/LeafHandler.ashx?n=pet&s=emm_epm0u_pte_swa_dpg&f=a

We can also see the number of rows and columns, and the column or attribute names themselves  of the dataset here:

##### Row and Column Info of Gas Prices Dataset - 1st Dataset

To explore the dimensions of the gas price dataframe/table, we can see it using the following:

In [236]:
print("Row and Column Count:", df_gas.shape)
print("Attribute/Column Names:", df_gas.columns)

Row and Column Count: (25, 2)
Attribute/Column Names: Index(['Back to Contents', 'Data 1: Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)'], dtype='object')


Based on the above output, we can see that there are 25 rows, and 2 columns. Additionally, the 2 column names (or attributes as sometimes called in Machine Learning) can be seen under the shape output. 

From this, we can notice that there's something bizarre with the column names. Namely, the first column looks like a navigation button the original spreadsheet itself, and the second column seems to be the title of the tab in the spreadsheet. We can confirm that this is the case using the following function:

In [237]:
df_gas

,Back to Contents,Data 1: Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)
0,Sourcekey,EMM_EPM0U_PTE_SWA_DPG
1,Date,Washington All Grades Conventional Retail Gaso...
2,2003-06-30 00:00:00,1.684
3,2004-06-30 00:00:00,1.991
4,2005-06-30 00:00:00,2.41
5,2006-06-30 00:00:00,2.751
6,2007-06-30 00:00:00,3.009
7,2008-06-30 00:00:00,3.453
8,2009-06-30 00:00:00,2.626
9,2010-06-30 00:00:00,3.058


The above output shows us that the first two rows are some sort of metadata, whether the title of the tab on the spreadsheet (Data 1: Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)) or a navigation button to go back to the first tab. This is also confirmed when opening the spreadsheet in Excel or Google Sheets.

The second row also seems to be something that can be changed, as it's information for a source key of the data, which we won't be needing for our use case.

We can also check to make sure that all the dates recorded are on the date 06-30 in which case we can drop the timestamp and the day and month, as the years are the only thing that changes.

##### Row and Column Info of Electric Vehicle Population Dataset - 2nd Dataset

We can first look at the shape of the dataset using the following code:

In [238]:
print("Row and Column Count:", df_EVs.shape)
print("Attribute/Column Names:", df_EVs.columns)

Row and Column Count: (279780, 16)
Attribute/Column Names: Index(['VIN (1-10)', 'County', 'City', 'State', 'Postal Code', 'Model Year',
       'Make', 'Model', 'Electric Vehicle Type',
       'Clean Alternative Fuel Vehicle (CAFV) Eligibility', 'Electric Range',
       'Legislative District', 'DOL Vehicle ID', 'Vehicle Location',
       'Electric Utility', '2020 Census Tract'],
      dtype='object')


To see the first five rows, we can use the following:

In [239]:
df_EVs.head()

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Legislative District,DOL Vehicle ID,Vehicle Location,Electric Utility,2020 Census Tract
0,JN1AZ0CP5C,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,7.0,153331706,POINT (-117.90454 48.54657),AVISTA CORP,5.306595e+10
1,JTMABABA7P,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0.0,15.0,253586308,POINT (-120.71847 46.55029),PACIFICORP,5.307700e+10
2,1N4AZ1CP1J,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,37.0,333135022,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
3,5UX43EU09S,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid Electric Vehicle (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,23.0,267525737,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC,5.303594e+10
4,3C3CFFGE5F,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,2.0,474468501,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,5.306701e+10


From the above table, we can see that there's a couple of things that can be changed during the data cleaning process.
- VIN numbers: could potentially be treated as the id column for these cars as a primary key, assuming there are no duplicate entries, which need to be checked first.
- Postal codes: should be stored as integers and not floats, as the 0 at the end of each postal code is unneccesary.
- Electric Vehicle Type: can be just Battery or Plug-in Hybrid etc.
- CAFV Elibility: we can just use boolean variables for "Eligible" or "Not Eligible", and use NULL/NaN in the cases where the eligibility is unknown.
- Electric Range: can also be an integer and the column name can be renamed to Electric Range in Miles
- Vehicle Location: I'm going to leave this alone for now, as we can possibly map these as coordinates later. I did some research, and turns out these are map coordinates.
- Electric Utility: The CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA) cells can be simplified to CITY OF SEATTLE/TACOMA.

Columns we can drop:
- Legislative District: is a column that can be dropped, as it isn't relevant to our data analysis right now.
- DOL Vehicle ID: can also be dropped, as there isn't much of a use case for this, but it depends on if it helps to prevent duplicate IDs in the dataset, which we would need to check before dropping this column.
- 2020 Census Tract: We can drop this column, as it doesn't provide much information.

### Data Cleaning

#### Gas Prices
We can start by cleaning the gas prices dataset, associated with df_gas.

In [240]:
df_gas.head()

,Back to Contents,Data 1: Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)
0,Sourcekey,EMM_EPM0U_PTE_SWA_DPG
1,Date,Washington All Grades Conventional Retail Gaso...
2,2003-06-30 00:00:00,1.684
3,2004-06-30 00:00:00,1.991
4,2005-06-30 00:00:00,2.41


We can assign the third row (including the header) as the new header. We can do this using the following, as the index of the third row is 2. Which is why we write header=2 as one of the parameters in the read_excel function call below.

In [241]:
df_gas = pd.read_excel("Gas_Prices.xls", sheet_name="Data 1", header=2)
df_gas.head()

,Date,Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)
0,2003-06-30,1.684
1,2004-06-30,1.991
2,2005-06-30,2.410
3,2006-06-30,2.751
4,2007-06-30,3.009


I also decided to make the second column's header name of "Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)" to "WA Retail Gasoline Prices (Dollars Per Gallon)" using the following code.

In [242]:
df_gas.rename(columns={'Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)': 'WA Retail Gas Prices (Dollars per Gallon)'}, inplace=True)
df_gas.head()

,Date,WA Retail Gas Prices (Dollars per Gallon)
0,2003-06-30,1.684
1,2004-06-30,1.991
2,2005-06-30,2.410
3,2006-06-30,2.751
4,2007-06-30,3.009


Now, if we'd like to get rid of the 06-30 part in all of the dates, we need to check to make sure that the year is the only value that changes. In order to check that, we can take a look at the column of dates:

In [243]:
df_gas['Date']

0    2003-06-30
1    2004-06-30
2    2005-06-30
3    2006-06-30
4    2007-06-30
5    2008-06-30
6    2009-06-30
7    2010-06-30
8    2011-06-30
9    2012-06-30
10   2013-06-30
11   2014-06-30
12   2015-06-30
13   2016-06-30
14   2017-06-30
15   2018-06-30
16   2019-06-30
17   2020-06-30
18   2021-06-30
19   2022-06-30
20   2023-06-30
21   2024-06-30
22   2025-06-30
Name: Date, dtype: datetime64[ns]

We can see that the column type is datetime as shown at the bottom of the above output. Additionally, the only values that change are the years, so we can effectively remove the month and days. We can use the following code to extract just the year:

In [244]:
df_gas['Date'] = df_gas['Date'].dt.year
df_gas.head()

,Date,WA Retail Gas Prices (Dollars per Gallon)
0,2003,1.684
1,2004,1.991
2,2005,2.410
3,2006,2.751
4,2007,3.009


Although the year can be used as an index, I will leave it alone because we will be using the year column to merge the two tables later.

#### Electric Vehicle Population Dataset Cleaning

We can also implement the changes I mentioned earlier in the EV Population Dataset.

In [245]:
df_EVs.head()

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Legislative District,DOL Vehicle ID,Vehicle Location,Electric Utility,2020 Census Tract
0,JN1AZ0CP5C,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,7.0,153331706,POINT (-117.90454 48.54657),AVISTA CORP,5.306595e+10
1,JTMABABA7P,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0.0,15.0,253586308,POINT (-120.71847 46.55029),PACIFICORP,5.307700e+10
2,1N4AZ1CP1J,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,37.0,333135022,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
3,5UX43EU09S,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid Electric Vehicle (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,23.0,267525737,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC,5.303594e+10
4,3C3CFFGE5F,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,2.0,474468501,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,5.306701e+10


First, we can drop the duplicate rows if there are any in this dataset.

In [246]:
og_len = df_EVs.shape[0]
df_EVs.drop_duplicates()
new_len = df_EVs.shape[0]
print("Original Length:", og_len)
print("New Length:", new_len)

Original Length: 279780
New Length: 279780


Turns out that there weren't any duplicates in the dataset, which is why the length of the dataset didn't change after dropping duplicates.

Now we can check if each VIN number and DOL Vehicle ID is unique.

In [247]:
VIN = df_EVs['VIN (1-10)']
VIN.nunique()
VIN.value_counts()

7SAYGDEE7P    1187
7SAYGDEE6P    1180
7SAYGDEEXP    1166
7SAYGDEE5P    1157
7SAYGDEE8P    1152
              ... 
1G1RC6E43E       1
YV4BR0CK5N       1
KMTGE4S11R       1
7JRH60FD0N       1
3GPKHURM9S       1
Name: VIN (1-10), Length: 17072, dtype: int64

In [248]:
df_EVs['DOL Vehicle ID'].nunique()

279780

Turns out that the VIN (1-10) column cannot be used as the id of this row since the unique entries are only about 17000 entries, which is much lower than the total of 279780 entries. So, we can drop the VIN (1-10) column, since it doesn't seem to be relevant to this exploration.

On the other hand, the DOL Vehicle ID column can be reassigned as the index column, as it seems that there are 279780 unique values, which is the same exact length as the original dataset. We can do this like so:

In [249]:
df_EVs.set_index('DOL Vehicle ID', inplace=True)
df_EVs.head()

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Legislative District,Vehicle Location,Electric Utility,2020 Census Tract
DOL Vehicle ID,,,,,,,,,,,,,,,
153331706,JN1AZ0CP5C,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,7.0,POINT (-117.90454 48.54657),AVISTA CORP,5.306595e+10
253586308,JTMABABA7P,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0.0,15.0,POINT (-120.71847 46.55029),PACIFICORP,5.307700e+10
333135022,1N4AZ1CP1J,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,37.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
267525737,5UX43EU09S,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid Electric Vehicle (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,23.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC,5.303594e+10
474468501,3C3CFFGE5F,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,2.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,5.306701e+10


Now, we can proceed to drop the columns that are unnecessary for this exploration.
- Legislative District
- VIN (1-10)
- 2020 Census Tract

In [250]:
df_EVs.drop(columns=['VIN (1-10)', 'Legislative District', '2020 Census Tract'], inplace=True)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0.0,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid Electric Vehicle (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


Next up, we can change the data saved in the Electric Vehicle Type column to just Plug-in Hybrid, Battery Electric, or NaN, but we first have to check the column values.

In [251]:
df_EVs.value_counts('Electric Vehicle Type')

Electric Vehicle Type
Battery Electric Vehicle (BEV)            223884
Plug-in Hybrid Electric Vehicle (PHEV)     55896
dtype: int64

Now that we've confirmed that there's only two types of values, we can replace them like so:

In [252]:
df_EVs['Electric Vehicle Type'].replace({'Plug-in Hybrid Electric Vehicle (PHEV)': 'Plug-in Hybrid (PHEV)', 'Battery Electric Vehicle (BEV)': 'Battery Electric (BEV)'}, inplace=True)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric (BEV),Eligibility unknown as battery range has not b...,0.0,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


We can also change the next column, which is Clean Alternative Fuel Vehicle (CAFV) Eligibility like so:
- Clean Alternative Fuel Vehicle Eligible -> True
- Eligibility Unknown... -> NaN or NULL
- Not Eligible -> False

First, let's check the values of the column:

In [253]:
df_EVs.value_counts('Clean Alternative Fuel Vehicle (CAFV) Eligibility')

Clean Alternative Fuel Vehicle (CAFV) Eligibility
Eligibility unknown as battery range has not been researched    177937
Clean Alternative Fuel Vehicle Eligible                          77495
Not eligible due to low battery range                            24348
dtype: int64

Now, we can replace those values accordingly:

In [254]:
df_EVs['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].replace({
'Eligibility unknown as battery range has not been researched': None,
'Clean Alternative Fuel Vehicle Eligible': True,
'Not eligible due to low battery range': False,
}, inplace=True)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric (BEV),True,73.0,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,0.0,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric (BEV),True,151.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric (BEV),True,87.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


A couple of code blocks ago, we also saw that the CAFV Eligibility column had some data that said: "Eligibility unknown as battery range has not been researched". For these columns you can see that the battery range is noted at 0.0 in the next column for electric range. We want to prevent averages getting dragged down, so we'd need to replace these 0.0s with np.nan.

I noticed when I used None the first time, the 0.0s were getting replaced with the values from the previous rows from above, which made the data have the wrong values. Upon further research, I learned that whenever you deal with numeric values, you need to use np.nan.

In [255]:
df_EVs['Electric Range'] = df_EVs['Electric Range'].replace(0.0, np.nan)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric (BEV),True,73.0,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,NaN,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric (BEV),True,151.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric (BEV),True,87.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


Next, we can change the postal code, and possibly the electric range to integers. In the case of electric ranges, we need to check if it has any numbers after the decimals to make sure that we won't lose any numerical precision.

In [256]:
# I got stuck at this point, so I had to use an LLM to check for rows where there are numbers with a decimal point and not null:
mask = (~np.isclose(df_EVs["Electric Range"] % 1, 0)) & pd.notna(df_EVs["Electric Range"])
df_EVs[mask]


,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,


Using the above output, we can see that there are no rows where there is a number after the decimal point. Therefore, we can use the following code to remove the zeroes from the end of each number:

In [257]:
df_EVs['Electric Range'] = df_EVs['Electric Range'].astype('Int64')
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric (BEV),True,73,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,<NA>,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric (BEV),True,151,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric (BEV),True,87,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


As a general principle, postal codes don't have decimal points, so we can remove the 0 from the end by casting each postal code, which is a float, to an int. We can check to see if we need to use 'Int64', based on how many null values there are in the dataset using the .info() function.

In [258]:
df_EVs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 279780 entries, 153331706 to 267728625
Data columns (total 12 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   County                                             279756 non-null  object 
 1   City                                               279756 non-null  object 
 2   State                                              279780 non-null  object 
 3   Postal Code                                        279756 non-null  float64
 4   Model Year                                         279780 non-null  int64  
 5   Make                                               279780 non-null  object 
 6   Model                                              279780 non-null  object 
 7   Electric Vehicle Type                              279780 non-null  object 
 8   Clean Alternative Fuel Vehicle (CAFV) Eligibility  101843 non-n

The above information shows us that there are a couple of entries that are null in the postal code as well, which means we ought to use 'Int64' casting.

In [259]:
df_EVs['Postal Code'] = df_EVs['Postal Code'].astype('Int64')
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114,2012,NISSAN,LEAF,Battery Electric (BEV),True,73,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,<NA>,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122,2018,NISSAN,LEAF,Battery Electric (BEV),True,151,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597,2015,FIAT,500,Battery Electric (BEV),True,87,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


Now that we've cleaned the datasets, we can merge/join them based on Year column in the df_gas and the Model Year in df_EVs. We can do this using the following:

In [266]:
df_gas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 2 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Date                                       23 non-null     int64  
 1   WA Retail Gas Prices (Dollars per Gallon)  23 non-null     float64
dtypes: float64(1), int64(1)
memory usage: 496.0 bytes


In [269]:
print(df_EVs.info())
print(df_gas.info())

df_merged = pd.merge(df_EVs, df_gas, how='left', on='Date', validate=None)
df_merged.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 279780 entries, 153331706 to 267728625
Data columns (total 12 columns):
 #   Column                                             Non-Null Count   Dtype 
---  ------                                             --------------   ----- 
 0   County                                             279756 non-null  object
 1   City                                               279756 non-null  object
 2   State                                              279780 non-null  object
 3   Postal Code                                        279756 non-null  Int64 
 4   Model Year                                         279780 non-null  int64 
 5   Make                                               279780 non-null  object
 6   Model                                              279780 non-null  object
 7   Electric Vehicle Type                              279780 non-null  object
 8   Clean Alternative Fuel Vehicle (CAFV) Eligibility  101843 non-null  objec

KeyError: 'Date'

### Exploratory Data Analysis (EDA)
Below is some python code I wrote as a preliminary exploration of the data using pandas.

First, I'm interested in seeing if the table loaded, and what the columns look like, so I'll display the first five rows of the dataset to see some sample rows.

In [233]:
df.head()

NameError: name 'df' is not defined

##### NOTES FOLLOWING THIS (REMOVE LATER):
Start with context and objective → shows business understanding
Show data cleaning & analysis → demonstrates practical skills
Visualize insights clearly → makes it engaging and readable
Conclude with insights & next steps → shows analytical thinking

##### LAYOUT FOR THIS PORTFOLIO
Title & Introduction
Objective / Questions
Data Loading
Data Cleaning / Preparation
Exploratory Data Analysis (EDA)
Visual Analysis (matplotlib + seaborn)
Key Insights
Conclusion / Next Steps

Optional: Dataset link, interactive plots, or extra analysis if relevant

### Notebook title (project name)
Short description of the project
Dataset source and context
Optional: an image or logo for visual appeal

### Objective / Questions
What you are trying to find out
Specific questions your analysis will answer

<h3 id="data-loading" style="color: mediumpurple;"> Data Loading </h3>
Load the dataset with pandas
Show basic info (head(), info(), describe())

### Data Cleaning / Preparation
Handle missing data
Filter rows/columns
Create new columns if needed
Group or aggregate data

### Exploratory Data Analysis (EDA)
Explore patterns in the data
Use pandas to summarize (groupby, value_counts)
Include matplotlib/seaborn plots for trends

### Visual Analysis
Use seaborn for cleaner, more advanced plots
Correlation heatmaps, scatterplots, boxplots, pairplots

### Key Insights
Summarize findings from analysis and plots
Use bullet points or markdown for clarity

### Conclusion / Next Steps
Wrap up project
Optional: suggest further analysis or improvements